feature/manager/src/main/java/com/kinandcarta/create/proxytoggle/manager/viewmodel/ProxyManagerViewModel.kt

```diff
@@ -0,0 +1,215 @@
+import androidx.compose.ui.text.input.ImeAction
+import androidx.compose.ui.text.input.KeyboardType
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.kinandcarta.create.proxytoggle.core.android.DeviceSettingsManager
+import com.kinandcarta.create.proxytoggle.core.android.ProxyValidator
+import com.kinandcarta.create.proxytoggle.core.android.ThemeSwitcher
+import com.kinandcarta.create.proxytoggle.core.model.Proxy
+import com.kinandcarta.create.proxytoggle.core.settings.AppSettings
+import com.kinandcarta.create.proxytoggle.manager.R
+import dagger.hilt.android.lifecycle.HiltViewModel
+import dagger.hilt.android.qualifiers.ApplicationContext
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@Suppress("TooManyFunctions")
+@HiltViewModel
+class ProxyManagerViewModel @Inject constructor(
+    @ApplicationContext private val context: Context,
+    private val deviceSettingsManager: DeviceSettingsManager,
+    private val proxyValidator: ProxyValidator,
+    private val appSettings: AppSettings,
+    private val themeSwitcher: ThemeSwitcher
+) : ViewModel() {
+
+    private var _uiState = mutableStateOf(
+        UiState(
+            darkTheme = themeSwitcher.isNightMode(),
+            proxyEnabled = false,
+            addressState = getInitialAddressState(),
+            portState = getInitialPortState()
+        )
+    )
+    val uiState: State<UiState> = _uiState
+
+    init {
+        viewModelScope.launch {
+            deviceSettingsManager.proxySetting.collect { proxy ->
+                if (proxy.isEnabled) {
+                    updateUiState {
+                        it.copy(
+                            proxyEnabled = true,
+                            addressState = it.addressState.copy(text = proxy.address, error = null),
+                            portState = it.portState.copy(text = proxy.port, error = null)
+                        )
+                    }
+                } else {
+                    updateUiState {
+                        it.copy(proxyEnabled = false)
+                    }
+                }
+            }
+        }
+    }
+
+    fun toggleProxy() {
+        if (deviceSettingsManager.proxySetting.value.isEnabled) {
+            deviceSettingsManager.disableProxy()
+        } else {
+            enableProxy()
+        }
+    }
+
+    fun toggleTheme() {
+        themeSwitcher.toggleTheme()
+        updateUiState { it.copy(darkTheme = themeSwitcher.isNightMode()) }
+    }
+
+    fun onAddressChanged(newText: String) {
+        val newTextFiltered = newText.filter { it.isDigit() || it == '.' }
+        updateUiState {
+            it.copy(addressState = it.addressState.copy(text = newTextFiltered))
+        }
+    }
+
+    fun onPortChanged(newText: String) {
+        val newTextFiltered = newText
+            .filter(Char::isDigit)
+            .take(ProxyValidator.MAX_PORT.toString().length)
```

Comment:
Maybe instead of taking the maximum port we can show an error saying "Max port is X"?

Reply:
Hm, this is part of a rudimentary check DURING typing, so that I can save the user from some easy mistakes. Same way I disallow letters. The final validation is always done by the viewmodel. I could revert to that, but I don't see how I could show this error, without making the errors dynamic, ie reacting to the user typing. It's quite the change!